dev_guide/app_tutorials/jenkins_pipeline.adoc

```diff
@@ -0,0 +1,219 @@
+          steps {
+            def rm = openshift.selector("dc", templateName).rollout()
+            // wait up to 5 minutes for the deployment to complete
+            timeout(5) {
+              openshift.selector("dc", templateName).related('pods').untilEach(1) {
+                return (it.object().status.phase == "Running")
+              }
+            }
+          }
+        }
+        stage('tag') {
+          steps {
+            // if everything else succeeded, tag the ${templateName}:latest image as ${templateName}:stage
+            // a pipeline build config for the staging environment can watch for the ${templateName}:stage
+            // image to change and then deploy it to the staging environment
+            openshift.tag("${templateName}:latest", "${templateName}:stage")
+          }
+        }
+      }
+    }
+  }
+}
+----
+
+[NOTE]
+====
+The above example has been written using the *declarative pipeline* style, but the older *scripted pipeline* style is also supported.
+====
+
+
+[[creating-the-pipeline]]
+
+== Creating the Pipeline
+
+Now we can create the build config in our OpenShift cluster with the following command:
+
+[source]
+----
+$ oc create -f nodejs-sample-pipeline.yaml
+----
+
+If you don't want to create your own file, you can use the sample from the Origin repository with the following command:
+
+[source]
+----
+$ oc create -f https://raw.githubusercontent.com/openshift/origin/master/examples/jenkins/pipeline/nodejs-sample-pipeline.yaml
+----
+
+For more information about the OpenShift DSL syntax used here see https://jenkins.io/doc/pipeline/steps/openshift-pipeline/[OpenShift Pipeline Steps].
+
+[[starting-the-pipeline]]
+
+== Starting the Pipeline
+
+We can now start the pipeline with the following command:
+
+[source]
+----
+$ oc start-build nodejs-sample-pipeline
+----
+
+[NOTE]
+====
+Alternatively, you could also start your pipeline via the OpenShift Web Console by navigating to the Builds -> Pipeline
+section, or by visiting the Jenkins Console, navigating to the Pipeline that you created, and clicking *Build Now*.
+====
+
+Once the pipeline has started you should see the following actions performed within your project:
+
+ * a job instance is created on the Jenkins server
+ * a node.js Jenkins slave pod is created
+ * any previously created resources with the *template=nodejs-mongodb-example* label will be deleted
+ * a new application (and all of it's associated resources) will be created from the *nodejs-mongodb-example* template
+ * a build will be started using the *nodejs-mongodb-example* build config
+ ** the pipeline will wait until the build has completed to trigger the next stage
+ * a deployment will be started using the *nodejs-mongodb-example* deployment config
+ ** the pipeline will wait until the deployment has completed to trigger the next stage
+ * if the build and deploy are successful, the *nodejs-mongodb-example:latest* image will be tagged as *nodejs-mongodb-example:stage*
```

Comment:
i'd divide this list into two sections:  things that always happen and things the pipeline specifically is doing.

Things that happen always:
1) job instance created
2) slave pod launched to run your pipeline (as long as your pipeline declares a node, so this one kinda muddies the water between a generic "what happens when you run a pipeline" action and "what the specific pipeline makes happen")
3) pipeline runs on slave
4) slave is deleted (if one was required for the pipeline)

things the pipeline does:
1) delete resources
2) create a new set of resources
3) run an openshift build to produce an app image
4) wait for build to complete
5) deploy the new app image
6) wait for deployment
7) tag the image